src/index.ts

```diff
@@ -1,132 +1,231 @@
+    wasmBufferFree(buf);
+    wasmBufferFree(result);
+    return json;
+}
+
+
+// Frees an allocated svm_codec buffer that was previously allocated and returned to caller by an api function
+export function wasmBufferFree(buf) : void {
+    return (codec.exports as any).wasm_free(buf);
+}
+
+// internal helper functions
+
+// Encodes binary provided binary data as a hex binary string (without an 0x prefix)
+function binToString(array: Uint8Array) : string {
+    let result = "";
+    for (const b of array) {
+        let s = b.toString(16);
+
+        // padding
+        if (s.length < 2) {
+            s = "0" + s;
+        }
+        result += s;
+    }
+    return result;
+}
+
+// Call an svm_codec function with the provided buffer. Returns result buffer.
+function call(funcName: string, buf) : Uint8Array {
+    return (codec.exports as any)[funcName](buf)
 }
 
-// Frees an allocated svm_codec buffer that was previously allocated by svm_codec
-export function wasmBufferFree(buf) {
-    return (codec!.exports as any).wasm_free(buf);
+// Allocates a svm_codec buffer with the provided byte length
+function wasmBufferAlloc(length: number) {
+    return (codec.exports as any).wasm_alloc(length)
 }
 
 // Returns the bytes length of a wasm_codec buffer
-export function wasmBufferLength(buf) {
-    return (codec!.exports as any).wasm_buffer_length(buf);
+function wasmBufferLength(buf) {
+    return (codec.exports as any).wasm_buffer_length(buf);
 }
 
 // Frees the data allocated in a svm_codec buffer
-export function wasmBufferDataPtr(buf) {
-    return (codec!.exports as any).wasm_buffer_data(buf);
+function wasmBufferDataPtr(buf) {
+    return (codec.exports as any).wasm_buffer_data(buf);
 }
 
 // Copies binary data from buf to an svm_codec memory buffer
-export function copyToWasmBufferData(buf, data) {
-    let ptr = wasmBufferDataPtr(buf);
-    let memory = (codec!.exports as any).memory.buffer;
-    let view = new Uint8Array(memory);
+function copyToWasmBufferData(buf, data) {
+    const ptr = wasmBufferDataPtr(buf);
+    const memory = (codec.exports as any).memory.buffer;
+    const view = new Uint8Array(memory);
     view.set([...data], ptr);
 }
 
 // Copies length bytes at an offset from buf to an svm_codec memory buffer
-export function wasmBufferDataSlice(buf, offset, length: number) {
-    let ptr = wasmBufferDataPtr(buf);
-    const memory = (codec!.exports as any).memory.buffer;
+function wasmBufferDataSlice(buf, offset, length: number) {
+    const ptr = wasmBufferDataPtr(buf);
+    const memory = (codec.exports as any).memory.buffer;
     const view = new Uint8Array(memory);
     return view.slice(ptr + offset, ptr + offset + length);
 }
 
 // Creates a wasm buffer that can be passed to an svm_codec instance methods from a provided json object
 // Returns the buffer
-export function newWasmBuffer(object: JSON) {
+function newWasmBuffer(object: any) {
     const objectStr = JSON.stringify(object);
```

Comment:
Is it safe enough regarding further buffer allocation?

E.G.
1. Call `newWasmBufer({ some: 23 });`
2. It will be stringified to `'{"some":23}'` (length: 11)
3. Then the buffer length will be allocated for 11 bytes
4. Then we need to update such data to be `{ some: 100500 }`
5. It will not fit the buffer size

Probably I don't understand how exactly svm works (I didn't dive into it yet).
But it doesn't seem to me that buffer is immutable since there are functions that do something with the buffer. So if we want to store some related data in the same memory allocation — we had to allocate it smarter by taking into account the size of the type. But it also will require to store it in a different way (not just stringify, and then easily JSON.parse), which is much more complex 🤔

Reply:
@YaronWittenstein ?

Reply:
All the buffers allocation are basically one-time operations.

* You have an object (JSON)
* Then you allocate its byte-size as a buffer
* Then you copy the JSON into that buffer
* Lastly you call the Wasm instance
* Now the buffer should be freed 

All this mechanism is internal of course  (not part of the exported API)